src/models/UndoRedo.ts

```diff
@@ -0,0 +1,210 @@
+//
+// redo:
+// Should be no-op
+// snapshots
+// ---------------------------------------
+// | initialState, "" | state0, "action" |
+// ---------------------------------------
+// currentSnapshot == 1
+// undoable should be 1, redoable should be 0
+//
+// commit
+// Set currentShapshot forward by 1.  preserve [0, currentSnapshot+1).  Append new state.
+// snapshots
+// ----------------------------------------------------------
+// | initialState, "" | state0, "action" | state1, "action" |
+// ----------------------------------------------------------
+// currentSnapshot == 2
+// undoable should be 1, redoable should be 0
+//
+// undo:
+// restore state at currentSnapshot.  Set currentSnapshot back by 1
+// snapshots
+// ----------------------------------------------------------
+// | initialState, "" | state0, "action" | state1, "action" |
+// ----------------------------------------------------------
+// currentSnapshot == 1
+// undoable should be 1, redoable should be 1
+//
+// commit
+// Set currentShapshot forward by 1.  preserve [0, currentSnapshot+1).  Append new state.
+// snapshots
+// ----------------------------------------------------------
+// | initialState, "" | state0, "action" | state2, "action" |
+// ----------------------------------------------------------
+// currentSnapshot == 2
+// undoable should be 1, redoable should be 0
+//
+// assume limit of 3
+// commit
+// rotate snapshots by 1 left.  override new state at currentShapshot-1.
+// snapshots
+// ----------------------------------------------------------
+// | state0, "action" | state2, "action" | state3, "action" |
+// ----------------------------------------------------------
+// currentSnapshot == 2
+// undoable should be 1, redoable should be 0
+
+export class UndoRedo extends Serializable<UndoRedo> {
+  metadataVersion: number = METADATA_VERSION;
+
+  // probably needs to be a pair, one of the state, and one describing what just happened.
+  stateSnapshots: [string, string][] = [];
+  currentShapshot = 0;
+  maxSnapshots = 10;
+
+  constructor(json: Partial<UndoRedo> = {}) {
+    super();
+    this.fromJson(json);
+  }
+
+  canUndo(): boolean {
+    return this.currentShapshot > 0;
+  }
+
+  undoString(): string {
+    // going to need to be a string that got us to this state.
+    return this.canUndo() ? this.stateSnapshots[this.currentShapshot][1] : "";
+  }
+
+  canRedo(): boolean {
+    return this.currentShapshot < this.stateSnapshots.length - 1;
+  }
+
+  redoString(): string {
+    // going to need to be the string of the next snapshot state.
+    return this.canRedo()
+      ? this.stateSnapshots[this.currentShapshot + 1][1]
+      : "";
+  }
+
+  reinitializeUndoRedo(show: Show): void {
+    this.stateSnapshots = [[JSON.stringify(show), ""]];
+    this.currentShapshot = 0;
+  }
+
+  createPlugin() {
+    return (store: Store<CalChartState>): void => {
+      this.stateSnapshots = [[JSON.stringify(store.state.show), ""]];
```

Comment:
Could we re-use code here?
```suggestion
      reinitializeUndoRedo(store.state.show);
```

Reply:
👍